session/session.go

```diff
@@ -46,19 +65,26 @@ func decryptCookie(cookie []byte, secret []byte) (dd []byte, err error) {
 	return
 }
 
-func DecryptSignedCookie(signed_cookie, secret_key_base, salt string) (session []byte, err error) {
-	cookie, err := url.QueryUnescape(signed_cookie)
+func DecryptSignedCookie(signedCookie, secretKeyBase, salt, signSalt string) (session []byte, err error) {
+	cookie, err := url.QueryUnescape(signedCookie)
 	if err != nil {
 		return
 	}
 
 	vectors := strings.SplitN(cookie, "--", 2)
+	if vectors[0] == "" || vectors[1] == "" {
+		return nil, errors.New("Invalid cookie!")
+	}
+	if verified := verifySign(vectors[0], vectors[1], secretKeyBase, signSalt); !verified {
+		return nil, errors.New("Not verified signature!")
```

Comment:
Same as above. Also I'd probably consider returning a sentinel error here, so the caller could easily distinguish between the input error and invalid signature cases.